### PR TITLE
CADC-8819 - Refactor skaha-wsproxy to remove duplicate code.

### PR DESCRIPTION
**Pull request by SharonGoliath**
Add some unit tests - 83% coverage.

### CODE REVIEW

File: skaha-wsproxy/conf/skaha_rewrite.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        :param log_fqn: Fully-qualified name for the log file.
+        :param max_size: maximum size of the cache
+        :param ttl: time to live of cache entries
+        """
+        self._log_fqn = log_fqn
+        self._log_file = open(log_fqn, 'a')
+        self._cache = TTLCache(maxsize=max_size, ttl=ttl)
+
+    def _build_url(self, segs, path, session_id, ip_address, params):
+        return None
```
Comment (andamian) [inline skaha-wsproxy/conf/skaha_rewrite.py]: If this is meant to be abstract shouldn't raise a notimplementederror here?

File: skaha-wsproxy/conf/skaha_rewrite.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                           '--kubeconfig=/root/kube/k8s-config',
+                           'get',
+                           'pod',
+                           '--selector=canfar-net-sessionID={}'.format(
+                               session_id),
+                           '--no-headers=true',
+                           '-o',
+                           'custom-columns=IPADDR:.status.podIP,'
+                           'DT:.metadata.deletionTimestamp']
+                command_string = ' '.join([str(elem) for elem in command])
```
Comment (andamian) [inline skaha-wsproxy/conf/skaha_rewrite.py]: Simple `' '.join(command)` should work instead, no? All the elements seem to be str already. Why is command an array if a str is needed at the end?

File: skaha-wsproxy/conf/skaha_rewrite.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                sys.stdout.write(response + '\n')
+            else:
+                self.log('INFO: End response: None')
+                sys.stdout.write('https://www.canfar.net/notfound.html\n')
+        except Exception as e:
+            tb = traceback.format_exc()
+            self.log('ERROR: unexpected: {}:{}'.format(str(e), tb))
+            sys.stdout.write('https://www.canfar.net/notfound.html\n')
+        except:
+            self.log('ERROR: unclassified error')
```
Comment (andamian) [inline skaha-wsproxy/conf/skaha_rewrite.py:177]: I'm not sure I understand the reason this is done this way. I would catch `Error` above which covers everything. With this implementation, errors like simple syntax errors, will be tricky to detect.

File: skaha-wsproxy/conf/skaha_rewrite.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            sys.stdout.write('https://www.canfar.net/notfound.html\n')
+        sys.stdout.flush()
+
+    def log(self, message):
+        self._log_file.write(time.ctime() + u' - ' + message.encode() + u'\n')
```
Comment (andamian) [inline skaha-wsproxy/conf/skaha_rewrite.py:182]: Could use `logging` maybe?

**Review by andamian — APPROVED** [review]

File: skaha-wsproxy/conf/skaha_rewrite.py (hunk 1 above)
Comment (SharonGoliath) [reply skaha-wsproxy/conf/skaha_rewrite.py]: It should, and now it does.

File: skaha-wsproxy/conf/skaha_rewrite.py (hunk 2 above)
Comment (SharonGoliath) [reply skaha-wsproxy/conf/skaha_rewrite.py]: Removed the array.

File: skaha-wsproxy/conf/skaha_rewrite.py (hunk 4 above)
Comment (SharonGoliath) [reply skaha-wsproxy/conf/skaha_rewrite.py:182]: Maybe. I don't know the logic behind what messages went to the log file, and what messages went to stdout, and I don't understand `logging` well enough to divide messaging that way without more time to research than I've got left on the task. I will if you think it's important though.

File: skaha-wsproxy/conf/skaha_rewrite.py (hunk 3 above)
Comment (SharonGoliath) [reply skaha-wsproxy/conf/skaha_rewrite.py:177]: The effect is to not print a stacktrace for KeyboardInterrupt. I don't if that was on purpose or not, though. I can check with Brian?

**Comment by brianmajor** [comment]
Just wanted to note how this has unique logging requirement.  stdout, stderr and stdin are used to communicate with apache, so nothing should be going to stdout for logging or error reporting.

**Pull request merged**